interfaces/builtin/dbus.go

```diff
@@ -0,0 +1,402 @@
+    peer=(label=###PLUG_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugAppArmor = `
+# allow snaps to introspect ###DBUS_NAME###. This allows us to
+# introspect other interfaces of the service (but not access them).
+dbus (send)
+    bus=###DBUS_BUS###
+    path=###DBUS_INTROSPECT_PATH###
+    interface=org.freedesktop.DBus.Introspectable
+    member=Introspect
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to ###DBUS_NAME###
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    interface=###DBUS_INTERFACE###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+
+# allow connected snaps to all interfaces via ###DBUS_PATH### (eg,
+# org.freedesktop.*, org.gtk.Application, etc) to allow full integration with
+# connected snaps.
+dbus (receive, send)
+    bus=###DBUS_BUS###
+    path=###DBUS_PATH###
+    peer=(label=###SLOT_SECURITY_TAGS###),
+`
+
+const dbusConnectedPlugSecComp = `
+getsockname
+recvmsg
+sendmsg
+sendto
+`
+
+type DbusInterface struct{}
+
+func (iface *DbusInterface) Name() string {
+	return "dbus"
+}
+
+// Obtain yaml-specified bus well-known name
+func (iface *DbusInterface) getAttribs(attribs map[string]interface{}) (string, string, error) {
+	bus := ""
+	name := ""
+	for attr := range attribs {
+		if attr != "bus" && attr != "name" {
+			return "", "", fmt.Errorf("unknown attribute '%s'", attr)
+		}
+
+		raw, ok := attribs[attr]
+		if !ok {
+			return "", "", fmt.Errorf("cannot find attribute %q", attr)
+		}
+		val, ok := raw.(string)
+		if !ok {
+			return "", "", fmt.Errorf("element %v for '%s' is not a string", raw, attr)
+		}
+
+		if attr == "bus" {
+			if val != "session" && val != "system" {
+				return "", "", fmt.Errorf("bus '%s' must be one of 'session' or 'system'", val)
+			}
+			bus = val
+		} else if attr == "name" {
+			err := interfaces.ValidateDBusBusName(val)
+			if err != nil {
+				return "", "", err
+			}
+			name = val
+		}
+	}
+
+	if bus == "" {
+		return "", "", fmt.Errorf("required attribute 'bus' not specified")
+	} else if name == "" {
+		return "", "", fmt.Errorf("required attribute 'name' not specified")
+	}
+
+	return bus, name, nil
+}
+
+// Determine AppArmor dbus abstraction to use based on bus
+func getAppArmorAbstraction(bus string) (string, error) {
+	var abstraction string
+	if bus == "system" {
+		abstraction = "dbus-strict"
+	} else if bus == "session" {
+		abstraction = "dbus-session-strict"
+	} else {
+		return "", fmt.Errorf("unknown abstraction for specified bus '%q'", bus)
+	}
+	return abstraction, nil
+}
+
+// Calculate individual snippet policy based on bus and name
+func getAppArmorSnippet(policy []byte, bus string, name string) []byte {
+	old := []byte("###DBUS_BUS###")
+	new := []byte(bus)
+	snippet := bytes.Replace(policy, old, new, -1)
+
+	old = []byte("###DBUS_NAME###")
+	new = []byte(name)
+	snippet = bytes.Replace(snippet, old, new, -1)
+
+	// convert name to AppArmor dbus path (eg 'org.foo' to '/org/foo{,/**}')
+	dot_re := regexp.MustCompile("\\.")
```

Comment:
I found that I couldn't get rid of dbus path entirely due to snaps using the org.freedesktop.* interfaces and so instead have:

```
139 # allow connected snaps to ###DBUS_NAME###                                      
140 dbus (receive, send)
141     bus=###DBUS_BUS###
142     interface=###DBUS_INTERFACE###
143     peer=(label=###SLOT_SECURITY_TAGS###),
144                                                                                 
145 # allow connected snaps to all interfaces via ###DBUS_PATH### (eg,              
146 # org.freedesktop.*, org.gtk.Application, etc) to allow full integration with   
147 # connected snaps.                                                              
dbus (receive, send)
    bus=###DBUS_BUS###
    path=###DBUS_PATH###
    peer=(label=###SLOT_SECURITY_TAGS###),  
```

This allows us to match by either interface (and security peer label) or path (and security peer label) and avoids the problems of trying to specify them both at the same time. In other words, still have the rule you mention that doesn't require guessing, but then have another rule that allows using the org.freedesktop.DBus interface in predictable ways.

 I discussed this with @tyhicks and we agreed on the approach.